core/src/main/java/org/apache/calcite/rel/rel2sql/SqlImplementor.java

```diff
@@ -2244,6 +2254,11 @@ && hasAliasUsedInHavingClause()) {
           // Avoid losing the distinct attribute of inner aggregate.
           return !hasNestedAgg || Aggregate.isNotGrandTotal(agg);
         }
+        if (relInput instanceof LogicalProject
+            && relInput.getInput(0) instanceof LogicalProject
+            && clauses.contains(Clause.QUALIFY)) {
+          return true;
```

Comment:
it feels like most of the changes in this PR can simply by solved by changing the order of evaluation of QUALIFY to after SELECT. can you share the impact of that change?